src/token/token.service.ts

```diff
@@ -104,7 +109,28 @@ export class TokenService {
       100
     );
   }
-  async mktCap() {
-    return Number(await this.totalCirculation()) * (await this.getPrice());
+  async marketDetails() {
+    const details = (await this.cacheManager.get('marketCap')) as MarketDetails;
+    if (details) {
+      return details;
+    } else {
+      try {
+        const price = await this.getPrice();
+        const circulatingSupply = await this.totalCirculation();
+        const marketCap = price * circulatingSupply;
+        const marketDetails = {
+          price,
+          circulatingSupply,
+          marketCap,
+        };
+        await this.cacheManager.set('marketCap', marketDetails as any, 30);
+        return marketDetails;
+      } catch (error) {
+        throw new InternalServerErrorException(
```

Comment:
What happens when it expires and we get here? I would prefer to use the last known good value but I don't know if that matches with what we need for the site

Reply:
This block should be reachable only if one of the calculations fail. In that scenario I think we should throw instead of displaying wrong numbers. Also, if I understood correctly from the cache-manager docs, once the TTL elapses it will return undefined for that key anyways